### ISSUE TITLE
fusion transactions are free "because they are done within a zero fee requirement" - can this be abused? why not?

### ISSUE THREAD

**Issue by SoreGums** [turtlecoin/meta#132]
A few of us had a chat today about Fusion transactions, log below, also attached is a PDF file with the same and the link to the actual start of the chat: 
https://discordapp.com/channels/388915017187328002/388928195585638400/509943652689248257

[TRTL_Network-dev_general-fusions.pdf](https://github.com/turtlecoin/meta/files/2560421/TRTL_Network-dev_general-fusions.pdf)

> 
> TRTL Network
> dev_general
> 08-Nov-2018
> 
> SoreGums 08-Nov-18 11:14 AM
> general topic: fusion transactions are free "because they are done within a zero fee requirement" - can this be abused? why not?
> 
> zpalmtree 08-Nov-18 11:16 AM
> well the requirements make it so you can't spam them once your wallet is optimized without making it unoptimized again, which costs fees
> also, miners will always take the highest fee per byte txs
> so if they hang out in the pool that's ok
> 
> SoreGums 08-Nov-18 11:18 AM
> sure there is that part - however if manually constructing tx, can these requirements be abused is what i mean. i get that zedwallet is sensible
> 
> zpalmtree 08-Nov-18 11:21 AM
> @SoreGums no, they are network requirements
> 
> SoreGums 08-Nov-18 11:22 AM
> yeah, i haven't looked at them, so i'm just asking can they be taken advantage of - is it purley tied to something that can't be manipulated because it is what it is? like outputs have to be sourced from x where x is only 1 possible value related to an address etc (edited)
> 
> iburnmycd 08-Nov-18 11:25 AM
> There’s a required ratio of ins to outs (edited)
> 
> SoreGums 08-Nov-18 11:26 AM
> yeah i've seen that one
> that's not address specific
> 
> zpalmtree 08-Nov-18 11:27 AM
> tx size <= 28,000 bytes or so, input:output ratio >= 4, min of 12 inputs
> idk why addresses would matter
> not that we can say anything about addresses from a tx
> 
> SoreGums 08-Nov-18 11:29 AM
> so acn just manually create tx that fit that and not be a fusion?
> 
> zpalmtree 08-Nov-18 11:29 AM
> if it fits the requirements and it has zero fee it's a fusion
> 
> SoreGums 08-Nov-18 11:29 AM
> ie send to a different adddress
> 
> zpalmtree 08-Nov-18 11:29 AM
> you don't have to send to your own address
> 
> SoreGums 08-Nov-18 11:29 AM
> thus flood - ofc like u said, miners can just pick higher fee tx (edited)
> 
> zpalmtree 08-Nov-18 11:29 AM
> it's still a valid fusion
> 
> SoreGums 08-Nov-18 11:30 AM
> when i said fusion, i meant as the idea of consolidating one's own wallet (edited)
> sure it is valid fusion, just asking about the intent and if there is a posible abuse scenario, that could degrade the networks performance (edited)
> now with added adjectives ^^
> 
> iburnmycd 08-Nov-18 11:32 AM
> The only way it degrades the networks performance is mishandling by the daemon.
> 
> zpalmtree 08-Nov-18 11:32 AM
> it's a net decrease in outputs so it's not like you can send fusion back and forth without paying fees at some point you can't send anymore
> same address or different it doesn't matter
> 
> SoreGums 08-Nov-18 11:33 AM
> useful for an exchange to avoid fees then ;)
> i'd put in the effort.
> 😛
> 
> iburnmycd 08-Nov-18 11:34 AM
> Assuming they have the inputs to always send out as fusions.
> And in the amounts they need.
> 
> SoreGums 08-Nov-18 11:34 AM
> manual tx assembly, can make own output amounts and go where needed?
> fusion to self to get req outputs to go external (edited)
> 
> iburnmycd 08-Nov-18 11:34 AM
> I’m not sure tbh if you can have multiple transfers in a fusion.
> 
> RockSteady (TRTL) 08-Nov-18 11:34 AM
> that would be possible
> 
> iburnmycd 08-Nov-18 11:35 AM
> I suppose so
> So yeah. Sure.
> 
> RockSteady (TRTL) 08-Nov-18 11:35 AM
> Do you mean optimize for an extremely low target, and then optimize that with a larger target and a different recipient
> Let's say I have 100MM and I optimize for a target of like 1000
> 
> SoreGums 08-Nov-18 11:36 AM
> i just mean an exachnge wallet has a lot of outputs to play with.
> thus once the works is done, the process of sending 0 fee tx is possible
> cause the network has a way
> 
> RockSteady (TRTL) 08-Nov-18 11:36 AM
> then I try to send 100MM knowing I can't and it will force optimize
> then I set the recipient instead of myself in the fusion
> sounds plausible
> 
> iburnmycd 08-Nov-18 11:37 AM
> You can’t optimize for a lower target... that’s a deoptimization... it’s splitting a key.
> 
> RockSteady (TRTL) 08-Nov-18 11:37 AM
> Yes you can
> Ooze does it
> did/does
> you have a slider you can set for your optimization target
> 
> iburnmycd 08-Nov-18 11:38 AM
> If you have 100M key image. To get anything lower it’s a splitting of a key.
> 
> SoreGums 08-Nov-18 11:38 AM
> as long as the ratio is met, the tx is marked fusion. outputs to any needed inputs is doable
> 
> iburnmycd 08-Nov-18 11:38 AM
> That’s not “optimizing” in my mind. That’s making your wallet bigger. More space. Less optimization. Lol.
> 
> RockSteady (TRTL) 08-Nov-18 11:38 AM
> I'm just saying, I suspect that you can take a fully optimized wallet, and then probably split some images without spending, and then you can further send it to someone other than yourself in a 0 fee tx as a fusion, right?
> let's assume for the sake of argument we can mix whatever denomination
> 
> SoreGums 08-Nov-18 11:39 AM
> an exchange wouldn't optimise to a big number is also my point, if the goal is free tx then that is the goal from the start and wouldn't involve optimise to large number single output
> 
> RockSteady (TRTL) 08-Nov-18 11:40 AM
> the outbound transaction has to be some type of fusion, doesn't have to be bigger
> im just trying to figure out ways to never pay a network fee
> 
> SoreGums 08-Nov-18 11:40 AM
> so starting with a single 100M keyimage isn't really on point 😛
> 
> RockSteady (TRTL) 08-Nov-18 11:40 AM
> I just pulled a number out of my ass
> I think that it may be possible to only transaction through fusions, that's the point im trying to make
> or mostly through fusions
> that's if you can set a recipient other than yourself during fusion
> 
> iburnmycd 08-Nov-18 11:41 AM
> I don’t think you can break it down without “spending” on the chain.
> 
> RockSteady (TRTL) 08-Nov-18 11:41 AM
> How? just optimize for a lower target
> is something preventing you from doing that
> 
> iburnmycd 08-Nov-18 11:41 AM
> Which happens on chain.
> The ratio.
> 
> RockSteady (TRTL) 08-Nov-18 11:42 AM
> ooze did it just fine
> 
> iburnmycd 08-Nov-18 11:42 AM
> Of in’s to outs
> 
> RockSteady (TRTL) 08-Nov-18 11:42 AM
> did it subtract a fee when going down and not up or what
> 
> SoreGums 08-Nov-18 11:42 AM
> 1x 100M keyimage plus 19 other small 0.1 keyimages to>> 5x rounded number keyimages > ratio met, done (edited)
>  upvote 1
> 
> RockSteady (TRTL) 08-Nov-18 11:42 AM
> ah forgot about the ratio, so what if you pad it with dust? or near dust?
> bound to have some if you're an exchange
> 
> iburnmycd 08-Nov-18 11:42 AM
> Probably. Sure.
> 
> RockSteady (TRTL) 08-Nov-18 11:43 AM
> I wonder if this is something that could be an issue
> 
> iburnmycd 08-Nov-18 11:43 AM
> But you’re still stuck only creating so many.
> You’d have to break it down repeatedly on chain
> 
> SoreGums 08-Nov-18 11:44 AM
> so i'm getting the feeling that it is possible - however won't degrade the network in a noticable way because the miners have a say in selecting tx and will pick higher fee tx over lower fee and thus regular tx will still be processed.
> 
> The only issue is if the mempool gets too big and delays the relaying of legit tx
> 
> iburnmycd 08-Nov-18 11:44 AM
> Honestly... down to 0.1
> And the. Just fuse those to get back up again.
> 
> RockSteady (TRTL) 08-Nov-18 11:44 AM
> i dont think fusion is beholden to the same fee issue
> is it?
> otherwise they'd just never process
> 
> iburnmycd 08-Nov-18 11:44 AM
> All of your inputs down to a collection is 0.1s and that’s it. Lol.
> 
> SoreGums 08-Nov-18 11:44 AM
> z said the fee impacts seletion
> i am assumed this was true
> 
> RockSteady (TRTL) 08-Nov-18 11:45 AM
> I wonder if fusion is the same
> 
> SoreGums 08-Nov-18 11:45 AM
> if all there are is fusion tx, then the miner will just add them, no worries, they are small after all
> 
> iburnmycd 08-Nov-18 11:45 AM
> What’s the question again?
> Oh. Yeah. the daemon by default compares transactions in the mempool.
> 
> RockSteady (TRTL) 08-Nov-18 11:45 AM
> because if the network is validating tx's without any special rules for fusion, then a fusion would have a fee below threshold and reject it, so im suggesting that fusions probably aren't governed into blocks with the same fee rules as a regular tx
> 
> iburnmycd 08-Nov-18 11:46 AM
> Fee. Then size. Then time in pool IIRC
>  upvote 1
> 
> RockSteady (TRTL) 08-Nov-18 11:46 AM
> what allows a fusion into the pool without a fee
> 
> iburnmycd 08-Nov-18 11:46 AM
> The ratio. Lol.
> 
> SoreGums 08-Nov-18 11:46 AM
> is\fusion_ < gets tx into the mempool i guess (edited)
> 
> iburnmycd 08-Nov-18 11:46 AM
> If the ratio is met... the fee can be 0
> Otherwise. Gtfo
> 
> RockSteady (TRTL) 08-Nov-18 11:47 AM
> :t_think:
> 
> SoreGums 08-Nov-18 11:47 AM
> ratio + byte size + is_fusion + fee = OK to the mempool (edited)
> 
> iburnmycd 08-Nov-18 11:47 AM
> Is_fusion is a derivative
> It’s not a “state”
> Ratio + size + fee = is fusion
> 
> SoreGums 08-Nov-18 11:48 AM
> oh really? ok then - interesting gotcha for the protocol then to keep track of when devs do other implementations (edited)
> ahh i don't see fusion tx for some reason
> yeah cause you need to derive it by this...
> 
> zpalmtree 08-Nov-18 11:49 AM
> the threshold is a spook
> it just says don't optimize key images that are >= this amount
> you can't increase outputs
> 
> iburnmycd 08-Nov-18 11:49 AM
> But. You can only send for free if the ratio and size are met
> We could just solve this problem.
> And charge for fusions. Lol.
> 
> SoreGums 08-Nov-18 11:50 AM
> haha 😛
> 
> iburnmycd 08-Nov-18 11:50 AM
> Problem solved. :gavel:
> Lol.
> 
> SoreGums 08-Nov-18 11:51 AM
> so if a wallet uses a remote node, are they paying the remote node convience fee for fusion tx?
> 
> iburnmycd 08-Nov-18 11:51 AM
> No
> I didn’t implement the node fee on fusions... because the policy said it was free. Lol
> I wanted to. I soooooo wanted to. Lol
> Honestly. Rock is right.
> Wen Bulletproofs
> 
> zpalmtree 08-Nov-18 11:54 AM
> you don't need to do anything smart to work out if it's a fusion - let the daemon do that. if it's 0 fee and not a coinbase, it's a fusion
> 
> iburnmycd 08-Nov-18 11:55 AM
> They are arguing constructing fusions differently in an effort to game the daemon and the network into allowing “normal” transactions that look like fusions to save the 0.1 fee (edited)
> My ancestors are beaming with pride.
>  t_ayy 1
> 
> iburnmycd 08-Nov-18 12:00 PM
> Picking say 12x 0.1 inputs and one 100M image — creating 4 outputs - say 3x 25,000,000.4 outputs going to yourself and a 25M output going to someone else.
> 
> zpalmtree 08-Nov-18 12:01 PM
> oh sure you can do that
> who cares
> lol
> 
> iburnmycd 08-Nov-18 12:02 PM
> Shit. No.
> It wouldn’t be 3x 25M.4
> 
> zpalmtree 08-Nov-18 12:02 PM
> reminds me of this
>  
> network is more optimized, i don't give a damn
> 
> iburnmycd 08-Nov-18 12:02 PM
> It’d be 4x 25m plus 3x 0.4
> So 13:7
> Or. 4x25 + 1x1 + 1x.2
> 13:6
> 
> zpalmtree 08-Nov-18 12:05 PM
> >actually wasting your time doing the math
> 
> iburnmycd 08-Nov-18 12:05 PM
> Because they have to be pretty numbers. Lol
> 
> zpalmtree 08-Nov-18 12:05 PM
> :brainlet:
> 
> iburnmycd 08-Nov-18 12:05 PM
> I know.
> > Someone on the internet is wrong. (edited)
> Lol
> 
> SoreGums 08-Nov-18 12:08 PM
> i want to wrap this up and archive it in meta...
> 
> general topic: fusion transactions are free "because they are done within a zero fee requirement" - can this be abused? why not?
> 
> it can be taken advantage of, however it has minimal noticeable impact on the network's regular tx throughput.
> 
> BECAUSE: there are fixed requirements that must be met
> 1) ratio on inputs to outputs
> 2) size be smaller than n bytes
> 3) fee be 0 (biggest mitigator of toruble making)
> 
> given the above miners then will still preference regular tx with fees over these fee free tx to include them in blocks.
> perhaps there is an impact on the mempools load if a large number of tx are in the mempool, the rebutal to this is that given the current network characteristics not an issue and if in the future this could be an issue with increased tx volume by then the whole point will most likely be moot because tx amounts will have vanished entirely due to bulletproofs or other overall improvements to the design of the core network protocol.
> 
> iburnmycd 08-Nov-18 12:09 PM
> The permutations you’d need to have to break that down and avoid the fee for even a % of transactions...
> Just pay the 0.1 you cheap skate. Lol.
> You’ll spend more than that in CPU time and effort. Lol.
> 
> SoreGums 08-Nov-18 12:10 PM
> if I'm an exhange I'm not using turtle-service in the first place... so while I'm replacing that I could spend a few hours adding the coded need to get free tx 😛
> how's my summary? (edited)
> going to use DiscordChatExporter to archive this - it works good
> 
> RockSteady (TRTL) 08-Nov-18 12:11 PM
> just copy paste the whole convo
> ez pz
> 
> SoreGums 08-Nov-18 12:11 PM
> yeah i am - that's what it does, except nicely formated isntead of weird
> 
> RockSteady (TRTL) 08-Nov-18 12:11 PM
> ah yeah or chatexplorer
> 
> iburnmycd 08-Nov-18 12:11 PM
> There’s a limit to number of fusions per block as well IIRC
> 
> RockSteady (TRTL) 08-Nov-18 12:11 PM
> shit i hope so
> we'd be so fucked
> 
> iburnmycd 08-Nov-18 12:12 PM
> @zpalmtree Chief Justice. Come here. Lol
> 
> SoreGums 08-Nov-18 12:12 PM
> i thought that was a different chain/project - limit of fusion tx per block... (edited)
> 
> zpalmtree 08-Nov-18 12:13 PM
> I think i made that claim at one point because i only ever saw about 6 fusions per block - but that's probably due to the size requirements (edited)
> playing vidya
> 
> RockSteady (TRTL) 08-Nov-18 12:13 PM
> also playing vidya 😄
> 
> zpalmtree 08-Nov-18 12:13 PM
> I'm not sure why a limit is needed on fusions tho
> 
> RockSteady (TRTL) 08-Nov-18 12:13 PM
> figured out trtl clicker
> have clicked twice
> 
> zpalmtree 08-Nov-18 12:13 PM
> they're 0 fee so lowest priority and it's better to have big blocks clear them out faster
> 
> RockSteady (TRTL) 08-Nov-18 12:13 PM
> about to open my second pineapple farm franchise
> 
> SoreGums 08-Nov-18 12:14 PM
> derp ofc the limit is block size 😛
> 
> iburnmycd 08-Nov-18 12:14 PM
> And low priority.
> 
> SoreGums 08-Nov-18 12:14 PM
> turtle clicker, i opened console after 3 borg purchases
> 
> RockSteady (TRTL) 08-Nov-18 12:14 PM
> that's what i did before
> what im doing now is wayyyyy better
> no console
> 
> iburnmycd 08-Nov-18 12:15 PM
> Let me know when your exchange gets those slow boat transactions to users for free
> 
> RockSteady (TRTL) 08-Nov-18 12:15 PM
> no mods
> and it's foolproof
